File: vizGrimoireJS/scm-analysis.R
## Copyright (C) 2012, 2013 Bitergia
##
## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 3 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program; if not, write to the Free Software
## Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
##
## This file is a part of the vizGrimoire.R package
## http://vizgrimoire.bitergia.org/
##
## Analyze and extract metrics data gathered by CVSAnalY tool
## http://metricsgrimoire.github.com/CVSAnalY
##
## Authors:
##   Jesus M. Gonzalez-Barahona <jgb@bitergia.com>
##   Alvaro del Castillo <acs@bitergia.com>
##   Daniel Izquierdo Cortazar <dizquierdo@bitergia.com>
##
##
## Usage:
##  R --vanilla --args -d dbname < scm-analysis.R
## or
##  R CMD BATCH scm-analysis.R
##

library("vizgrimoire")
library("ISOweek")
options(stringsAsFactors = FALSE) # avoid merge factors for toJSON 

conf <- ConfFromOptParse()
SetDBChannel (database = conf$database, user = conf$dbuser, password = conf$dbpassword)

if (conf$granularity == 'years') { 
    period = 'year'
    nperiod = 365
} else if (conf$granularity == 'months') { 
    period = 'month'
    nperiod = 31
} else if (conf$granularity == 'weeks') { 
    period = 'week'
    nperiod = 7
} else if (conf$granularity == 'days'){ 
    period = 'day'
    nperiod = 1
} else {stop(paste("Incorrect period:",conf$granularity))}

# destination directory
destdir <- conf$destination

# multireport
reports=strsplit(conf$reports,",",fixed=TRUE)[[1]]

#########
#EVOLUTIONARY DATA
#########

# 1- Retrieving and 2- merging data
evol_data = GetSCMEvolutionaryData(period, conf$startdate, conf$enddate, conf$identities_db)

if ('companies' %in% reports) { 
    companies <- EvolCompanies(period, conf$startdate, conf$enddate)
    evol_data = merge(evol_data, companies, all = TRUE)
}
if ('countries' %in% reports) {
    countries <- EvolCountries(period, conf$startdate, conf$enddate)
    evol_data = merge(evol_data, countries, all = TRUE)
}

evol_data <- completePeriodIds(evol_data, conf$granularity, conf)
evol_data <- evol_data[order(evol_data$id), ]
evol_data[is.na(evol_data)] <- 0

# 3- Creating a JSON file 
createJSON (evol_data, paste(destdir,"/scm-evolutionary.json", sep=''))

#########
#STATIC DATA
#########

# 1- Retrieving information
static_data = GetSCMStaticData(period, conf$startdate, conf$enddate, conf$identities_db)
static_url <- StaticURL()
latest_activity7 = last_activity(7)
latest_activity14 = last_activity(14)
latest_activity30 = last_activity(30)
latest_activity60 = last_activity(60)
latest_activity90 = last_activity(90)
latest_activity180 = last_activity(180)
latest_activity365 = last_activity(365)
latest_activity730 = last_activity(730)

#Data for specific analysis
if ('companies' %in% reports){
	static_data_companies = evol_info_data_companies (conf$startdate, conf$enddate)
        static_data = merge(static_data, static_data_companies)
}
if ('countries' %in% reports){ 
	static_data_countries = evol_info_data_countries (conf$startdate, conf$enddate)
        static_data = merge(static_data, static_data_countries)
}
# 2- Merging information
static_data = merge(static_data, static_url)
static_data = merge(static_data, latest_activity7)
static_data = merge(static_data, latest_activity14)
static_data = merge(static_data, latest_activity30)
static_data = merge(static_data, latest_activity60)
static_data = merge(static_data, latest_activity90)
static_data = merge(static_data, latest_activity180)
static_data = merge(static_data, latest_activity365)
static_data = merge(static_data, latest_activity730)


# 3- Creating file with static data
createJSON (static_data, paste(destdir,"/scm-static.json", sep=''))


# Top authors

top_authors_data <- top_authors(conf$startdate, conf$enddate)
top_authors_data <- list()
top_authors_data[['authors.']] <- top_people(0, conf$startdate, conf$enddate, "author" , "-Bot" )
top_authors_data[['authors.last year']]<- top_people(365, conf$startdate, conf$enddate, "author", "-Bot")
top_authors_data[['authors.last month']]<- top_people(31, conf$startdate, conf$enddate, "author", "-Bot")
createJSON (top_authors_data, paste(destdir,"/scm-top.json", sep=''))

# Top files
top_files_modified_data = top_files_modified()

if ('companies' %in% reports) {

    companies  <- companies_name_wo_affs(c("-Bot", "-Individual", "-Unknown"), conf$startdate, conf$enddate)
    companies <- companies$name
    createJSON(companies, paste(destdir,"/scm-companies.json", sep=''))
	
    for (company in companies){
        company_name = paste("'", company, "'", sep='')
        company_aux = paste("", company, "", sep='')
        print (company_name)
	
        ######
        #Evolutionary data per company
        ######	
        # 1- Retrieving and merging info  
        evol_data = GetSCMEvolutionaryData(period, conf$startdate, conf$enddate, conf$identities_db, list("company", company_name))
        		
        evol_data <- completePeriodIds(evol_data, conf$granularity, conf)
        evol_data <- evol_data[order(evol_data$id), ]
        evol_data[is.na(evol_data)] <- 0
		
        # 2- Creation of JSON file
        createJSON(evol_data, paste(destdir,"/",company_aux,"-scm-evolutionary.json", sep=''))
				
        ########
        #Static data per company
        ########
        static_data <- GetSCMStaticData(period, conf$startdate, conf$enddate, conf$identities_db, list("company", company_name))

        createJSON(static_data, paste(destdir,"/",company_aux,"-scm-static.json", sep=''))
	
        top_authors <- company_top_authors(company_name, conf$startdate, conf$enddate)
        createJSON(top_authors, paste(destdir,"/",company_aux,"-scm-top-authors.json", sep=''))
        top_authors_2006 <- company_top_authors_year(company_name, 2006) 
        createJSON(top_authors_2006, paste(destdir,"/",company_aux,"-scm-top-authors_2006.json", sep=''))
        top_authors_2009 <- company_top_authors_year(company_name, 2009)
        createJSON(top_authors_2009, paste(destdir,"/",company_aux,"-scm-top-authors_2009.json", sep=''))
        top_authors_2012 <- company_top_authors_year(company_name, 2012)
        createJSON(top_authors_2012, paste(destdir,"/",company_aux,"-scm-top-authors_2012.json", sep=''))	
    }
}

if ('repositories' %in% reports) {
    repos  <- repos_name(conf$startdate, conf$enddate)
    repos <- repos$name
    limit = 30
    if (length(repos)<limit) limit = length(repos);
    repos <- repos[1:limit]
    createJSON(repos, paste(destdir,"/scm-repos.json", sep=''))
	
    for (repo in repos) {
        repo_name = paste("'", repo, "'", sep='')
        repo_aux = paste("", repo, "", sep='')
        print (repo_name)
        
        ###########
        #EVOLUTIONARY DATA
        ###########
        #1- Retrieving data
  
        evol_data = GetSCMEvolutionaryData(period, conf$startdate, conf$enddate, conf$identities_db, list("repository", repo_name))
        evol_data <- completePeriodIds(evol_data, conf$granularity, conf)
        evol_data <- evol_data[order(evol_data$id), ]
        evol_data[is.na(evol_data)] <- 0
        
        #3- Creating JSON
        createJSON(evol_data, paste(destdir, "/",repo_aux,"-scm-evolutionary.json", sep=''))
		
        ##########
        #STATIC DATA
        ##########
        # 1- Retrieving information
        static_data = GetSCMStaticData(period, conf$startdate, conf$enddate, conf$identities_db, list("repository", repo_name))

        #3- Creating JSON
        #static_info <- evol_info_data_repo(repo_name, period, conf$startdate, conf$enddate)
        createJSON(static_data, paste(destdir, "/",repo_aux,"-scm-static.json", sep=''))		
    }		
}

if ('countries' %in% reports) {
    countries  <- scm_countries_names(conf$identities_db,conf$startdate, conf$enddate)
    countries <- countries$name
    createJSON(countries, paste(destdir,"/scm-countries.json", sep=''))
	
    for (country in countries) {
        if (is.na(country)) next
        print (country)
        country_name = paste("'", country, "'", sep='')
        
        evol_data = GetSCMEvolutionaryData(period, conf$startdate, conf$enddate, conf$identities_db, list("country", country_name))
        # evol_data <- EvolCommits(period, conf$startdate, conf$enddate, conf$identities_db, country=country_name)
        evol_data <- completePeriodIds(evol_data, conf$granularity, conf)
        # evol_data <- evol_data[order(evol_data$id), ]
        # evol_data[is.na(evol_data)] <- 0
        
        createJSON (evol_data, paste(destdir, "/",country,"-scm-evolutionary.json",sep=''))
        
        # data <- scm_countries_static(conf$identities_db, country, conf$startdate, conf$enddate)
        static_data = GetSCMStaticData(period, conf$startdate, conf$enddate, conf$identities_db, list("country", country_name))
        createJSON (static_data, paste(destdir, "/",country,"-scm-static.json",sep=''))
    }
}

if ('people' %in% reports) {
    print ('Starting people analysis')
    people  <- GetPeopleListSCM(conf$startdate, conf$enddate)
    people = people$pid
    limit = 100
    if (length(people)<limit) limit = length(people);
    people = people[1:limit]
    createJSON(people, paste(destdir,"/scm-people.json", sep=''))
	
    for (upeople_id in people) {
        evol_data <- GetEvolPeopleSCM(upeople_id, period, 
                conf$startdate, conf$enddate)
        evol_data <- completePeriodIds(evol_data, conf$granularity, conf)
        evol_data[is.na(evol_data)] <- 0
        createJSON (evol_data, paste(destdir,"/people-",
                        upeople_id,"-scm-evolutionary.json", sep=''))
        static_data <- GetStaticPeopleSCM(upeople_id, 
                conf$startdate, conf$enddate)
        createJSON (static_data, paste(destdir,"/people-",
                        upeople_id,"-scm-static.json", sep=''))        
    }        
}

if ('companies-countries' %in% reports){
    companies  <- companies_name(conf$startdate, conf$enddate)
    companies <- companies$name
    for (company in companies){
        countries  <- scm_countries_names(conf$identities_db,conf$startdate, conf$enddate)
    countries <- countries$name
    for (country in countries) {
            company_name = paste(c("'", company, "'"), collapse='')
            company_aux = paste(c("", company, ""), collapse='')

            ###########
            if (is.na(country)) next
            print (paste(country, "<->", company))
            data <- scm_companies_countries_evol(conf$identities_db, company, country, nperiod, conf$startdate, conf$enddate)
            if (length(data) == 0) {
                data <- data.frame(id=numeric(0),commits=numeric(0),authors=numeric(0))
            }

            data = completeZeroPeriod(data, nperiod, conf$str_startdate, conf$str_enddate)
            data$week <- as.Date(conf$str_startdate) + data$id * nperiod
            data$date  <- toTextDate(GetYear(data$week), GetMonth(data$week)+1)
            data <- data[order(data$id), ]
            createJSON (data, paste("data/json/companycountry/",company,".",country,"-scm-evolutionary.json",sep=''))

            data <- scm_countries_static(conf$identities_db, country, conf$startdate, conf$enddate)
            createJSON (data, paste("data/json/companycountry/",company,".",country,"-scm-static.json",sep=''))

            #################


        }
    }
}

# Demographics
d <- new ("Demographics","scm",6)
people <- Aging(d)
people$age <- as.Date(conf$str_enddate) - as.Date(people$firstdate)
people$age[people$age < 0 ] <- 0
aux <- data.frame(people["id"], people["age"])
new <- list()
new[['date']] <- conf$str_enddate
new[['persons']] <- aux
createJSON (new, paste(c(destdir, "/scm-demographics-aging.json"), collapse=''))

newcomers <- Birth(d)
newcomers$age <- as.Date(conf$str_enddate) - as.Date(newcomers$firstdate)
newcomers$age[newcomers$age < 0 ] <- 0
aux <- data.frame(newcomers["id"], newcomers["age"])
new <- list()
new[['date']] <- conf$str_enddate
new[['persons']] <- aux
createJSON (new, paste(c(destdir, "/scm-demographics-birth.json"), collapse=''))

File: vizGrimoireJS/scm-analysis.py
#!/usr/bin/env python

# Copyright (C) 2014 Bitergia
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
# This file is a part of the vizGrimoire.R package
#
## Authors:
##   Jesus M. Gonzalez-Barahona <jgb@bitergia.com>
##   Alvaro del Castillo San Felix <acs@bitergia.com>
##   Daniel Izquierdo Cortazar <dizquierdo@bitergia.com>
#
# Usage:
#     PYTHONPATH=../vizgrimoire LANG= R_LIBS=../../r-lib ./mls-analysis.py 
#                                                -d acs_irc_automatortest_2388_2 -u root 
#                                                -i acs_cvsanaly_automatortest_2388 
#                                                -s 2010-01-01 -e 2014-01-20 
#                                                -o ../../../json -r people,repositories
#

import logging
from rpy2.robjects.packages import importr
import sys

isoweek = importr("ISOweek")
vizr = importr("vizgrimoire")

import GrimoireUtils, GrimoireSQL
from GrimoireUtils import dataFrame2Dict, createJSON, completePeriodIds
from GrimoireUtils import valRtoPython, read_options, getPeriod
import SCM

def aggData(period, startdate, enddate, identities_db, destdir):
    # data = dataFrame2Dict(vizr.GetSCMStaticData(period, startdate, enddate, identities_db))
    data = SCM.GetSCMStaticData(period, startdate, enddate, identities_db, None)
    agg = data
    static_url = SCM.StaticURL()
    agg = dict(agg.items() + static_url.items())

    if ('companies' in reports):
        data = SCM.evol_info_data_companies (startdate, enddate)
        agg = dict(agg.items() + data.items())

    if ('countries' in reports): 
        data = SCM.evol_info_data_countries (startdate, enddate)
        agg = dict(agg.items() + data.items())

    if ('domains' in reports):
        data = SCM.evol_info_data_domains (startdate, enddate)
        agg = dict(agg.items() + data.items())

    data = SCM.GetCodeCommunityStructure(period, startdate, enddate, identities_db)
    agg = dict(agg.items() + data.items())

    # TODO: repeated data
    # data = dataFrame2Dict(vizr.GetDiffCommitsDays(period, enddate, 365))
    # agg = dict(agg.items() + data.items())

    # Tendencies    
    for i in [7,30,365]:
        data = SCM.GetDiffCommitsDays(period, enddate, identities_db, i)
        agg = dict(agg.items() + data.items())
        data = SCM.GetDiffAuthorsDays(period, enddate, identities_db, i)
        agg = dict(agg.items() + data.items())
        data = SCM.GetDiffFilesDays(period, enddate, identities_db, i)
        agg = dict(agg.items() + data.items())
        data = SCM.GetDiffLinesDays(period, enddate, identities_db, i)
        agg = dict(agg.items() + data.items())

    # Last Activity: to be removed
    for i in [7,14,30,60,90,180,365,730]:
        data = SCM.last_activity(i)
        agg = dict(agg.items() + data.items())

    # Fields with wrong data in R
    skip_fields = ['percentage_removed_lines_30','percentage_added_lines_30','diff_netadded_lines_30','diff_netremoved_lines_30']
    createJSON (agg, destdir+"/scm-static.json", True, skip_fields)

def tsData(period, startdate, enddate, identities_db, destdir, granularity, conf):
#    data = vizr.GetSCMEvolutionaryData(period, startdate, enddate, identities_db)
#    evol_data = completePeriodIds(dataFrame2Dict(data))
    data = SCM.GetSCMEvolutionaryData(period, startdate, enddate, identities_db, None)
    evol_data = completePeriodIds(data)

    if ('companies' in reports) :
        data = SCM.EvolCompanies(period, startdate, enddate)
        evol_data = dict(evol_data.items() + completePeriodIds(data).items())

    if ('countries' in reports) :
        data = SCM.EvolCountries(period, startdate, enddate)
        evol_data = dict(evol_data.items() + completePeriodIds(data).items())

    if ('domains' in reports) :
        data = SCM.EvolDomains(period, startdate, enddate)
        evol_data = dict(evol_data.items() + completePeriodIds(data).items())
 
    createJSON (evol_data, destdir+"/scm-evolutionary.json")

def peopleData(period, startdate, enddate, identities_db, destdir, top_authors_data):
    top = top_authors_data['authors.']["id"]
    top += top_authors_data['authors.last year']["id"]
    top += top_authors_data['authors.last month']["id"]
    # remove duplicates
    people = list(set(top))
    # the order is not the same than in R json
    createJSON(people, destdir+"/scm-people.json", False)

    for upeople_id in people :
        evol_data = SCM.GetEvolPeopleSCM(upeople_id, period, startdate, enddate)
        evol_data = completePeriodIds(evol_data)
        createJSON (evol_data, destdir+"/people-"+str(upeople_id)+"-scm-evolutionary.json")

        agg = SCM.GetStaticPeopleSCM(upeople_id,  startdate, enddate)
        createJSON (agg, destdir+"/people-"+str(upeople_id)+"-scm-static.json")

    pass

def reposData(period, startdate, enddate, identities_db, destdir, conf):
    # repos  = dataFrame2Dict(vizr.repos_name(startdate, enddate))
    repos  = SCM.repos_name(startdate, enddate)
    repos = repos['name']
    if not isinstance(repos, (list)): 
        repos = [repos]
        createJSON(repos, destdir+"/scm-repos.json", False)
    else:
        createJSON(repos, destdir+"/scm-repos.json")

    for repo in repos :
        repo_name = "'"+ repo+ "'"
        print (repo_name)

        evol_data = SCM.GetSCMEvolutionaryData(period, startdate, enddate, identities_db, ["repository", repo_name])
        evol_data = completePeriodIds(evol_data)
        createJSON(evol_data, destdir+"/"+repo+"-scm-rep-evolutionary.json")

        agg = SCM.GetSCMStaticData(period, startdate, enddate, identities_db, ["repository", repo_name])
        createJSON(agg, destdir+"/"+repo+"-scm-rep-static.json")

def companiesData(period, startdate, enddate, identities_db, destdir, bots, npeople):
    companies  = SCM.companies_name_wo_affs(bots, startdate, enddate)
    companies = companies['name']
    createJSON(companies, destdir+"/scm-companies.json")

    for company in companies:
        company_name = "'"+ company+ "'"
        print (company_name)

        evol_data = SCM.GetSCMEvolutionaryData(period, startdate, enddate, identities_db, ["company", company_name])
        evol_data = completePeriodIds(evol_data)
        createJSON(evol_data, destdir+"/"+company+"-scm-com-evolutionary.json")

        agg = SCM.GetSCMStaticData(period, startdate, enddate, identities_db, ["company", company_name])
        createJSON(agg, destdir+"/"+company+"-scm-com-static.json")

        top_authors = SCM.company_top_authors(company_name, startdate, enddate, npeople)
        createJSON(top_authors, destdir+"/"+company+"-scm-com-top-authors.json", False)

        for i in [2006,2009,2012]:
            data = SCM.company_top_authors_year(company_name, i, npeople)
            createJSON(data, destdir+"/"+company+"-scm-top-authors_"+str(i)+".json", False)

    commits =  SCM.GetCommitsSummaryCompanies(period, startdate, enddate, opts.identities_db, 10)
    createJSON (commits, destdir+"/scm-companies-commits-summary.json")

def countriesData(period, startdate, enddate, identities_db, destdir):
    countries  = SCM.scm_countries_names(identities_db,startdate, enddate)
    countries = countries['name']
    createJSON(countries, destdir+"/scm-countries.json")

    for country in countries:
        print (country)
        country_name = "'"+country+"'"

        evol_data = SCM.GetSCMEvolutionaryData(period, startdate, enddate, identities_db, ["country", country_name])
        evol_data = completePeriodIds(evol_data)
        createJSON (evol_data, destdir+"/"+country+"-scm-cou-evolutionary.json")

        agg = SCM.GetSCMStaticData(period, startdate, enddate, identities_db, ["country", country_name])
        createJSON (agg, destdir+"/"+country+"-scm-cou-static.json")

def domainsData(period, startdate, enddate, identities_db, destdir):
    domains = SCM.scm_domains_names(identities_db,startdate, enddate)
    domains = domains['name']
    createJSON(domains, destdir+"/scm-domains.json")
    # Some R ts are wrong
    bad_R_json_domains = ['gerrit','gmx','emsenhuber','bitergia']

    for domain in domains :
        domain_name = "'"+domain+"'"
        print (domain_name)

        evol_data = SCM.GetSCMEvolutionaryData(period, startdate, enddate, identities_db, ["domain", domain_name])
        evol_data = completePeriodIds(evol_data)
        if domain in bad_R_json_domains:
            createJSON(evol_data, destdir+"/"+domain+"-scm-dom-evolutionary.json", False)
        else:
            createJSON(evol_data, destdir+"/"+domain+"-scm-dom-evolutionary.json")

        agg = SCM.GetSCMStaticData(period, startdate, enddate, identities_db, ["domain", domain_name])
        createJSON(agg, destdir+ "/"+domain+"-scm-dom-static.json")


def companies_countriesData(period, startdate, enddate, identities_db, destdir):
    companies = SCM.companies_name(startdate, enddate)
    companies = companies['name']
    for company in companies:
        company_name = "'"+company+ "'"
        countries  = SCM.scm_countries_names(identities_db,startdate, enddate)
        countries = countries['name']
        for country in countries :
            print (country, "=>", company)
            data = SCM.scm_companies_countries_evol(identities_db, company, country, nperiod, startdate, enddate)
            data = completePeriodIds(data)
            createJSON (data, destdir + "/"+company+"_"+country+"-scm-evolutionary.json", False)

            # Not implemented in original R
            # data = vizr.scm_countries_static(identities_db, country, startdate, enddate)
            # createJSON (dataFrame2Dict(data), destdir + "/"+company+"_"+country+"-scm-static.json", False)

def topData(period, startdate, enddate, identities_db, destdir, bots, npeople):
    top_authors_data =  {}
    top_authors_data['authors.'] = SCM.top_people(0, startdate, enddate, "author" , "" , npeople)
    top_authors_data['authors.last year']= SCM.top_people(365, startdate, enddate, "author", "", npeople)
    top_authors_data['authors.last month']= SCM.top_people(31, startdate, enddate, "author", "", npeople)
    createJSON (top_authors_data, destdir+"/scm-top.json")

    # Top files
    top_files_modified_data = SCM.top_files_modified()

    return top_authors_data

def microStudies(enddate, destdir):
    # Studies implemented in R

    # Demographics
    vizr.ReportDemographicsAgingSCM(enddate, destdir)
    vizr.ReportDemographicsBirthSCM(enddate, destdir)


if __name__ == '__main__':
    logging.basicConfig(level=logging.INFO,format='%(asctime)s %(message)s')
    logging.info("Starting SCM data source analysis")
    opts = read_options()
    period = getPeriod(opts.granularity)
    nperiod = getPeriod(opts.granularity, True)
    reports = opts.reports.split(",")
    # filtered bots

    bots = ["-Bot", "-Individual", "-Unknown"]
    # TODO: hack because VizR library needs. Fix in lib in future
    startdate = "'"+opts.startdate+"'"
    enddate = "'"+opts.enddate+"'"

    # Working at the same time with VizR and VizPy yet
    vizr.SetDBChannel (database=opts.dbname, user=opts.dbuser, password=opts.dbpassword)
    GrimoireSQL.SetDBChannel (database=opts.dbname, user=opts.dbuser, password=opts.dbpassword)

    aggData(period, startdate, enddate, opts.identities_db, opts.destdir)
    tsData (period, startdate, enddate, opts.identities_db, opts.destdir, opts.granularity, opts)

    if ('domains' in reports):
        domainsData (period, startdate, enddate, opts.identities_db, opts.destdir)
    if ('repositories' in reports):
        reposData (period, startdate, enddate, opts.identities_db, opts.destdir, opts)
    if ('countries' in reports):
        countriesData (period, startdate, enddate, opts.identities_db, opts.destdir)
    if ('companies' in reports):
        companiesData (period, startdate, enddate, opts.identities_db, opts.destdir, bots, opts.npeople)
    if ('companies-countries' in reports):
        companies_countriesData (period, startdate, enddate, opts.identities_db, opts.destdir)
    # pretty slow!
    top = topData(period, startdate, enddate, opts.identities_db, opts.destdir, bots, opts.npeople)
    if ('people' in reports):
        peopleData (period, startdate, enddate, opts.identities_db, opts.destdir, top)
    microStudies(opts.enddate, opts.destdir)

    logging.info("SCM data source analysis OK")
